chore: added typescript definitions (#207)
vuelidate/formvuelatte: docs/3.x/src/guide/typescript.md

---
sidebarDepth: 3
---

# TypeScript Support <Badge text="3.3" type="warning" vertical="middle" />

> Formvuelate has typescript support for its public API. This page is an overview of the level of TypeScript support.

Reviewer: ```suggestion
FormVueLate has typescript support for its public API. This page is an overview of the level of TypeScript support.
```

## Component Types

Both `SchemaForm` and `SchemaWizard` have their properties and slots typed, to benefit from this you should use [volar](https://github.com/johnsoncodehk/volar) to enable template TypeScript support for components.

Furthermore the returned component created by `SchemaFormFactory` is also typed and will provide feedback if you pass incorrect type to any of its props.

> In addition to component types, Formvuelate also exposes various types you can generally in your project to provide type safety.

Reviewer: ```suggestion
In addition to component types, FormVueLate also exposes other types you can use in your project to provide type safety.
```

> Here is an example using `FormArraySchema` and `FormObjectSchema` to make it easier for you to create compatible schemas:

Reviewer: ```suggestion
For example, to create typed compatible schemas we can use `FormArraySchema` and `FormObjectSchema` as follows:
```

```ts

> import { SchemaForm, SchemaFormFactory, FormArraySchema } from "formvuelate";

Reviewer: I wonder if we should do something like `from formvuelate/types` so that user's imports are clearer

Author: It's common to keep types under the root namespace of the pkg. TypeScript users should be able to differentiate the difference between actual exports (useSchemaForm) and type exports (Interfaces, Types, etc..) so the separation is not necessary.

TypeScript won't let them use Type imports as values by mistake. Does this address your concern?

export default defineComponent({
  name: "App",
  components: {
    SchemaForm
  },
  setup() {
    const arraySchema = ref<FormArraySchema>([
      {
        component: FormText,
        model: 'firstName'
      },
      {
        model: 'lastName',
        component: FormText
      }
    ])

    const objectSchema = ref<FormObjectSchema>({
      firstName: {
        component: FormText
      },
      lastName: {
        component: FormText
      },
    })

    return {
      schema
    }
  }
})
```

## Plugin API Types

> Formvuelate also exposes some types to make it easier for you to create compatible plugins:

Reviewer: ```suggestion
FormVueLate also exposes the `PluginFunction` type to make it easier for you to create compatible plugins:
```
